src/main/java/org/broadinstitute/hellbender/tools/reference/ReferenceSequenceTable.java

```diff
@@ -0,0 +1,420 @@
+    }
+
+    private List<String> generateColumnNames() {
+        List<String> columns = new ArrayList<>();
+        columns.add(MD5_COLUMN_NAME);
+        columns.add(LENGTH_COLUMN_NAME);
+        for (GATKPath path : referenceSources.keySet()) {
+            columns.add(getReferenceColumnName(path));
+        }
+        return columns;
+    }
+
+    public List<String> getColumnNames() {
+        return columnNames;
+    }
+
+    /**
+     * Given a GATKPath, return the name of the file as a String.
+     *
+     * TODO: may need to add functionality to uniquify reference names
+     *
+     * @param reference The path to a reference.
+     * @return the name of the reference as a String.
+     */
+    public static String getReferenceColumnName(GATKPath reference){
+        return reference.toPath().getFileName().toString();
+    }
+
+    public Map<String, Integer> getColumnIndices() {
+        return columnIndices;
+    }
+
+    /**
+     * Construct 2 tables of references: one keyed by MD5, one keyed by sequence name.
+     */
+    public void build() {
+        tableByMD5 = new LinkedHashMap<>();
+        tableBySequenceName = new LinkedHashMap<>();
+
+        for (Map.Entry<GATKPath, ReferenceDataSource> entry : referenceSources.entrySet()) {
+            SAMSequenceDictionary dictionary = entry.getValue().getSequenceDictionary();
+            for (SAMSequenceRecord record : dictionary.getSequences()) {
+                String name = record.getSequenceName();
+                int length = record.getSequenceLength();
+                String md5 = getMD5ForRecord(record, entry.getKey());
+                GATKPath reference = entry.getKey();
+                TableEntry newEntry = new TableEntry(getReferenceColumnName(reference), name);
+                TableRow newRow = null;
+
+                // map each MD5 to List of TableEntry objects containing length, md5, and name
+                if (!tableByMD5.containsKey(md5)) {
+                    newRow = new TableRow(md5, length, new ArrayList<>(referenceSources.keySet()));
+                    tableByMD5.put(md5, newRow);
+                }
+                tableByMD5.get(md5).add(newEntry);
+
+                if(!tableBySequenceName.containsKey(name)){
+                    Set<TableRow> rowSet = new HashSet<>();
+                    tableBySequenceName.put(name, rowSet);
+                }
+                tableBySequenceName.get(name).add(tableByMD5.get(md5));
+            }
+        }
+        tableBuilt = true;
+    }
+
+    public Set<String> getAllSequenceNames(){
+        if(!tableBuilt){
+            throw new IllegalStateException("Table has not been built yet.");
+        }
+
+        return tableBySequenceName.keySet();
+    }
+
+    /**
+     * Given an MD5, returns its corresponding row
+     *
+     * @param md5 The MD5 as a String
+     * @return the corresponding TableRow from the tableByMD5
+     */
+    public TableRow queryByMD5(String md5){
+        if(!tableBuilt){
```

Comment:
Since you repeat this line in a bunch of places. Can you make a helper method `checkTableStatus()` or something like that and pull out all these `if(!tableBuilt)` statements to call it for readability.